src/jit.c

```diff
@@ -0,0 +1,1557 @@
+    emit4(state, 0);
+#elif defined(__aarch64__)
+    switch (code) {
+    case 0x84: /* BEQ */
+        code = COND_EQ;
+        break;
+    case 0x85: /* BNE */
+        code = COND_NE;
+        break;
+    case 0x8c: /* BLT */
+        code = COND_LT;
+        break;
+    case 0x8d: /* BGE */
+        code = COND_GE;
+        break;
+    case 0x82: /* BLTU */
+        code = COND_LO;
+        break;
+    case 0x83: /* BGEU */
+        code = COND_HS;
+        break;
+    default:
+        __UNREACHABLE;
+        break;
+    }
+    emit_a64(state, BR_Bcond | (0 << 5) | code);
+#endif
+}
+
+/* Load [src + offset] into dst */
+static inline void emit_load(struct jit_state *state,
+                             enum operand_size size,
+                             int src,
+                             int dst,
+                             int32_t offset)
+{
+#if defined(__x86_64__)
+    if (size == S8 || size == S16) {
+        /* movzx */
+        emit1(state, 0x0f);
+        emit1(state, size == S8 ? 0xb6 : 0xb7);
+    } else if (size == S32) {
+        /* mov */
+        emit1(state, 0x8b);
+    }
+
+    emit_modrm_and_displacement(state, dst, src, offset);
+#elif defined(__aarch64__)
+    switch (size) {
+    case S8:
+        emit_loadstore_imm(state, LS_LDRB, dst, src, offset);
+        break;
+    case S16:
+        emit_loadstore_imm(state, LS_LDRH, dst, src, offset);
+        break;
+    case S32:
+        emit_loadstore_imm(state, LS_LDRW, dst, src, offset);
+        break;
+    default:
+        __UNREACHABLE;
+        break;
+    }
+#endif
+}
+
+static inline void emit_load_sext(struct jit_state *state,
+                                  enum operand_size size,
+                                  int src,
+                                  int dst,
+                                  int32_t offset)
+{
+#if defined(__x86_64__)
+    if (size == S8 || size == S16) {
+        /* movsx */
+        emit1(state, 0x0f);
+        emit1(state, size == S8 ? 0xbe : 0xbf);
+    } else if (size == S32) {
+        emit_basic_rex(state, 1, dst, src);
+        emit1(state, 0x63);
+    }
+
+    emit_modrm_and_displacement(state, dst, src, offset);
+#elif defined(__aarch64__)
+    switch (size) {
+    case S8:
+        emit_loadstore_imm(state, LS_LDRSBW, dst, src, offset);
+        break;
+    case S16:
+        emit_loadstore_imm(state, LS_LDRSHW, dst, src, offset);
+        break;
+    default:
+        __UNREACHABLE;
+        break;
+    }
+#endif
+}
+
+/* Load sign-extended immediate into register */
+static inline void emit_load_imm(struct jit_state *state, int dst, int64_t imm)
+{
+#if defined(__x86_64__)
+    if (imm >= INT32_MIN && imm <= INT32_MAX)
+        emit_alu64_imm32(state, 0xc7, 0, dst, imm);
+    else {
+        /* movabs $imm, dst */
+        emit_basic_rex(state, 1, 0, dst);
+        emit1(state, 0xb8 | (dst & 7));
+        emit8(state, imm);
+    }
+#elif defined(__aarch64__)
+    if (imm >= INT32_MIN && imm <= INT32_MAX)
```

Comment:
Rewrite as `if ((int32_t)imm == imm)`, which is concise and should be as fast or faster than the original, depending on the compiler optimizations.